Join index rebuild of scheduler to assetmanager

Alaschinger opened the pull request:
Builds on PR #4120

- The repopulate function from the scheduler service has been removed.
- The index rebuild of scheduled events will now be proceessed as a part of the asset manager rebuild.
- A new SchedulerIndexData object will be created, selecting the relevant data from the scheduler database table.
- Within the rebuild of the assetmanager the scheduler data is checked for the given mediapackage id.
- If the event is a scheduled event it will by extended by the scheduler data, if not it will be processed as before.

Alaschinger committed:
Join index rebuild of scheduler to assetmanager

The repopulate function from the scheduler service has been removed. The index rebuild of scheduled events will now be proceessed as a part of the asset manager rebuild. Therefor a new SchedulerIndexData Object will be created, selecting the relevant data from the scheduler database table. Within the rebuild of the assetmanager the scheduler data is checked for the given mediapackage id. If the event is a scheduled event it will by extended by the scheduler data, if not it will be processed as before.

mtneug commented on etc/custom.properties:
```diff
@@ -113,16 +113,16 @@ org.opencastproject.download.directory=${org.opencastproject.storage.dir}/downlo
-#org.opencastproject.db.jdbc.pass=dbpassword
+org.opencastproject.db.jdbc.url=jdbc:mariadb://localhost/opencast?useMysqlMetadata=true
+org.opencastproject.db.jdbc.user=opencast
+org.opencastproject.db.jdbc.pass=dbpassword
```
This probably needs to be reverted?

Alaschinger replied:
Oh. Yes, this should be reverted. It was left from debugging.

lkiesow commented on etc/custom.properties:
```diff
@@ -113,16 +113,16 @@ org.opencastproject.download.directory=${org.opencastproject.storage.dir}/downlo
-#org.opencastproject.db.jdbc.url=jdbc:mariadb://localhost/opencast?useMysqlMetadata=true
-#org.opencastproject.db.jdbc.user=opencast
-#org.opencastproject.db.jdbc.pass=dbpassword
+org.opencastproject.db.jdbc.url=jdbc:mariadb://localhost/opencast?useMysqlMetadata=true
+org.opencastproject.db.jdbc.user=opencast
+org.opencastproject.db.jdbc.pass=dbpassword
```
Probably not, what you wanted to commit.

lkiesow commented on modules/asset-manager-impl/pom.xml:
```diff
@@ -66,6 +66,11 @@
       <artifactId>opencast-authorization-manager</artifactId>
       <version>${project.version}</version>
     </dependency>
+    <dependency>
+      <groupId>org.opencastproject</groupId>
+      <artifactId>opencast-scheduler-api</artifactId>
+      <version>${project.version}</version>
+    </dependency>
```
We should think about possible side effects here.
Thinking on potential for circular dependencies

Alaschinger replied:
I could easily remove this dependency. It was only used to get the string value of the the recording state 'unknown', which isn't expected to change.

mtneug commented:
This can still go into 13.0. Could you update the PR base branch to `r/13.x`?

Arnei commented:
Index rebuild is throwing an error if there never was a scheduled event. Unsure on how to best handle this.

```
2022-12-05 08:50:36,928 | WARN  | (Slf4JLogger:100) - Error: 1146-42S02: Table 'opencast.oc_scheduled_extended_event' doesn't exist
2022-12-05 08:50:36,931 | ERROR | (IndexEndpoint:163) - Repopulating the index failed
javax.persistence.PersistenceException: Exception [EclipseLink-4002] (Eclipse Persistence Services - 2.7.8.v20201217-ecdf3c32c4): org.eclipse.persistence.exceptions.DatabaseException
Internal Exception: java.sql.SQLSyntaxErrorException: (conn=4) Table 'opencast.oc_scheduled_extended_event' doesn't exist
Error Code: 1146
Call: SELECT mediapackage_id, capture_agent_id, recording_state, presenters, start_date, end_date FROM oc_scheduled_extended_event
Query: ResultSetMappingQuery(name="SchedulerIndexData.getAll" sql="SELECT mediapackage_id, capture_agent_id, recording_state, presenters, start_date, end_date FROM oc_scheduled_extended_event")
	at org.eclipse.persistence.internal.jpa.QueryImpl.getDetailedException(QueryImpl.java:391) ~[?:?]
	at org.eclipse.persistence.internal.jpa.QueryImpl.executeReadQuery(QueryImpl.java:264) ~[?:?]
	at org.eclipse.persistence.internal.jpa.QueryImpl.getResultList(QueryImpl.java:482) ~[?:?]
	at org.opencastproject.db.Queries$TypedQueriesBase.lambda$findAll$4(Queries.java:150) ~[?:?]
	at org.opencastproject.db.DBSessionImpl.execChecked(DBSessionImpl.java:79) ~[?:?]
	at org.opencastproject.db.DBSessionImpl.exec(DBSessionImpl.java:66) ~[?:?]
	at org.opencastproject.assetmanager.impl.persistence.Database.getSchedulerIndexData(Database.java:366) ~[?:?]
[...]
```

Arnei committed:
Catch error when querying non-existing db table

Using a NamedNativeQuery from an unrelated entity seems to not cause JPA to create the table that was queried.
Therefore, the query can cause an error if the table has not been created yet (e.g. because there has
not been an event scheduled yet). This commit catches that error. It might make more sense to relate
the entity "SchedulerIndexData" to "ExtendedEventDto" to create the table, but I am unsure on how
to do that.

gregorydlogan commented:
Looking at [Database:366](https://github.com/opencast/opencast/blob/a7bccdd47f91ebbc147a6533693a96c0ca82d8b3/modules/asset-manager-impl/src/main/java/org/opencastproject/assetmanager/impl/persistence/Database.java#L366), I'm not *entirely* sure what `insidePersistenceContextCheck` is supposed to be doing, but I do know that [the SnapshotDTO](https://github.com/opencast/opencast/blob/a7bccdd47f91ebbc147a6533693a96c0ca82d8b3/modules/asset-manager-impl/src/main/java/org/opencastproject/assetmanager/impl/persistence/SnapshotDto.java#L165) is the only place it's used upstream.  And what the check in `Database` is doing is ensuring that the DTO is being used within a JPA context.  Looking at the diff, you're touching DTO fields outside of an obvious JPA context.  Try that?

Arnei commented:
For posterity: I've talked with a Greg a bit and in the end he suggested that `getSchedulerIndexData()` should maybe not live where it lives right now, but ultimately didn't arrive at a solution.

KatrinIhler commented:
Talked to Julian and as there are no current plans to do work on this, I'm gonna close this for now.
